Composability of gateway plugin

dankelleher opened the pull request:
This PR brings composability (the ability to chain plugins) to the Civic Gateway plugin.

Specifically, the plugin registrar now accepts a "predecessor" plugin program id. If set, the plugin "inherits" the voter weight from the predecessor plugin. If not set, the plugin inherits the weight from the deposited governance tokens.

Note, the latter requires a slight change in core spl-gov to allow deposits when using a plugin (this is currently disabled). Until this change is made, this PR includes a built .so artifact of spl-governance that enables this (spl_governance_allow_deposits.so).

dankelleher committed:
Composability of gateway plugin

SebastianBor commented on programs/gateway/src/instructions/update_voter_weight_record.rs:
```diff
@@ -55,3 +66,26 @@ pub fn update_voter_weight_record(
+    match registrar.predecessor_plugin_registrar {
+        None => {
+            // If there is no predecessor plugin registrar, then the input account must be a TokenOwnerRecordV2
+            let parse_result: core::result::Result<TokenOwnerRecordV2, ProgramError> = get_account_data(&registrar.governance_program_id, input_account);
```
It's better to use `get_token_owner_record_data_for_realm_and_governing_mint()` here because
1) Deserialisation is backward compatible with `V1` accounts
2) It explicitly validates `TokenOwnerRecord` relationship to `Registrar`  primary key (`governance_program_id`, `realm`, `governing_token_mint`). You could even make it more explicit and create a function `get_token_owner_record_data_for_registrar()`

 https://github.com/solana-labs/solana-program-library/blob/3d3b32dcf6b64bfe609600735d067ebe735ed36d/governance/program/src/state/token_owner_record.rs#L354

dankelleher replied:
✅

SebastianBor commented on programs/gateway/src/instructions/update_voter_weight_record.rs:
```diff
@@ -55,3 +66,26 @@ pub fn update_voter_weight_record(
+            // If there is no predecessor plugin registrar, then the input account must be a TokenOwnerRecordV2
+            let parse_result: core::result::Result<TokenOwnerRecordV2, ProgramError> = get_account_data(&registrar.governance_program_id, input_account);
+            match parse_result {
+                Ok(token_owner_record) => token_owner_record.governing_token_deposit_amount,
```
Assert the deserialised `TokenOwnerRecord.governing_token_owner` matches `VoterWeightRecord. TokenOwnerRecord.governing_token_owner`

dankelleher replied:
✅

SebastianBor commented on programs/gateway/src/instructions/update_voter_weight_record.rs:
```diff
@@ -55,3 +66,26 @@ pub fn update_voter_weight_record(
+            let parse_result: core::result::Result<TokenOwnerRecordV2, ProgramError> = get_account_data(&registrar.governance_program_id, input_account);
+            match parse_result {
+                Ok(token_owner_record) => token_owner_record.governing_token_deposit_amount,
+                Err(_) => DEFAULT_VOTE_WEIGHT, // TODO should probably be an error
```
Yes, let's panic here

dankelleher replied:
✅

SebastianBor commented on programs/gateway/src/instructions/update_voter_weight_record.rs:
```diff
@@ -55,3 +66,26 @@ pub fn update_voter_weight_record(
+            }
+        }
+        Some(predecessor) => {
+            let parse_result: core::result::Result<VoterWeightRecord, ProgramError> = get_account_data(&predecessor, input_account);
```
We have to assert the deserialised `VoterWeightRecord` relationship with `Registrar.realm` and `Registrar.governing_token_mint`
nit: We can make it explicit and create a function `get_voter_weight_record_data_for_previous_plugin_and_registrar()`

dankelleher replied:
✅

SebastianBor commented on programs/gateway/src/instructions/update_voter_weight_record.rs:
```diff
@@ -55,3 +66,26 @@ pub fn update_voter_weight_record(
+        Some(predecessor) => {
+            let parse_result: core::result::Result<VoterWeightRecord, ProgramError> = get_account_data(&predecessor, input_account);
+            match parse_result {
+                Ok(predecessor_voter_weight_record) => predecessor_voter_weight_record.voter_weight,
```
Assert the deserialised previous plugin `VoterWeightRecord.governing_token_owner` matches our `VoterWeightRecord. TokenOwnerRecord.governing_token_owner`

dankelleher replied:
✅

SebastianBor commented on programs/gateway/src/instructions/update_voter_weight_record.rs:
```diff
@@ -55,3 +66,26 @@ pub fn update_voter_weight_record(
+            let parse_result: core::result::Result<VoterWeightRecord, ProgramError> = get_account_data(&predecessor, input_account);
+            match parse_result {
+                Ok(predecessor_voter_weight_record) => predecessor_voter_weight_record.voter_weight,
+                Err(_) => DEFAULT_VOTE_WEIGHT, // TODO should probably be an error
```
Yes, let's panic here

dankelleher replied:
✅

SebastianBor commented on programs/gateway/src/state/registrar.rs:
```diff
@@ -25,6 +25,10 @@ pub struct Registrar {
+    
+    /// If the plugin is one in a sequence, this is the previous plugin
+    /// If set, then update_voter_weight_record will expect a voter_weight_record owned by this registrar
+    pub predecessor_plugin_registrar: Option<Pubkey>,
```
Should it be `previous_voter_weight_plugin` to make it more explicit what plugin type we are expecting here?

dankelleher replied:
I renamed it - but kept the "registrar" to make it clear it should be the pubkey of a Registrar struct (when I think plugin, I think the plugin program Id)

SebastianBor commented on programs/gateway/src/state/registrar.rs:
```diff
@@ -25,6 +25,10 @@ pub struct Registrar {
+    
+    /// If the plugin is one in a sequence, this is the previous plugin
+    /// If set, then update_voter_weight_record will expect a voter_weight_record owned by this registrar
+    pub predecessor_plugin_registrar: Option<Pubkey>,
```
We would also need an instruction to configure both `previous_voter_weight_plugin` and `gatekeeper_network` for existing `Registrar`

dankelleher replied:
✅

SebastianBor commented on programs/gateway/src/instructions/update_voter_weight_record.rs:
```diff
@@ -41,6 +50,8 @@ pub fn update_voter_weight_record(
         &ctx.accounts.registrar.gatekeeper_network,
         None
     ).or(Err(error!(GatewayError::InvalidGatewayToken)))?;
+    
```
We don't need `voter_weight_action` and `target` params for the instruction because the `civic` plugin provided weight is not specific to any action or target

dankelleher replied:
✅

SebastianBor commented on programs/gateway/src/instructions/update_voter_weight_record.rs:
```diff
@@ -41,6 +50,8 @@ pub fn update_voter_weight_record(
         None
     ).or(Err(error!(GatewayError::InvalidGatewayToken)))?;
+    
+    let voter_weight = extract_input_voter_weight(&ctx.accounts.input_voting_weight.to_account_info(), &ctx.accounts.registrar);
```
When we have a previous plugin then  the data from its `VoteWeightRecord` (`voter_weight`, `weight_action`, `weight_action_target`, `reserved`) must be copied to our `VoteWeightRecord`

`voter_weight_expiry` should be set to `max`(`current_slot`, `previousVoteWeightRecord. voter_weight_expiry`) to either make the record valid only for the current slot or as long as the previous `VoteWeightRecord`
However the previous record expiry can't be longer then civic `gateway_token` validity if it has any expiration date

dankelleher replied:
✅

SebastianBor commented on programs/gateway/src/instructions/update_voter_weight_record.rs:
```diff
@@ -41,6 +50,8 @@ pub fn update_voter_weight_record(
         None
     ).or(Err(error!(GatewayError::InvalidGatewayToken)))?;
+    
+    let voter_weight = extract_input_voter_weight(&ctx.accounts.input_voting_weight.to_account_info(), &ctx.accounts.registrar);
```
If there is no previous plugin then let's explicitly set all fields on `VoteWeightRecord` to avoid stale state (ex. copied from another plugin when it was enabled and then disabled)

dankelleher replied:
✅

SebastianBor commented on programs/gateway/src/instructions/update_voter_weight_record.rs:
```diff
@@ -12,12 +15,18 @@ pub struct UpdateVoterWeightRecord<'info> {
+    /// depending on whether the registrar includes a predecessor or not
+    /// CHECK: Checked in the code depending on the registrar
+    #[account()]
+    pub input_voting_weight: UncheckedAccount<'info>,
```
I'm thinking maybe it should always be `VoterWeightRecord` to simplify plugins code.
We could expose `VoterWeightRecord` from `spl-governance` instead of using `TokenOwnerRecord`

dankelleher replied:
I'd be up for that - for now I have implemented a generic trait to simplify the code but it could be made simpler if we unify the two types.

SebastianBor reviewed (commented):
In order to support plugins stacking the upstream plugins must ensure no `vote_weight` reverse action can be performed within the `VoterWeightRecord` expiry time.
For example `Withdraw` instruction can't be executed before `voter_weight_expiry` to prevent the following attack

1) `upstream_plugin.deposit()` -> updates `VoterWeightRecord`
2) `civic_plugin.update_voter_weight_record()` -> updates civic `VoterWeightRecord`
3) `upstream_plugin.withdraw()` -> unlocks tokens and invalidates its own `VoterWeightRecord`
4) `spl-gov.cast_vote()` -> uses civic `VoterWeightRecord`

Using this pattern an attacker could repeat 1-4 indefinitely for different accounts using the same governance tokens 

For the reason above also `spl-gov` would have to be updated to support this temp. lock on tokens deposits

dankelleher committed:
Composability of gateway plugin - PR comments addressed - More tests TODO

dankelleher committed:
Composability of gateway plugin - Add tests for predecessor plugins, and bugfixes.

dankelleher committed:
Formatting, and removed voter weight action and target from the update voter weight instruction

dankelleher committed:
Renamed previous plugin registrar to previous plugin id

dankelleher committed:
v0.16.0-beta1

dankelleher committed:
Gateway Plugin: New program id supporting composability

dankelleher committed:
Gateway Plugin: New program id supporting composability

dankelleher committed:
Gateway Plugin: New program id and client lib supporting composability

dankelleher committed:
Gateway Plugin: Update Registrar instruction

dankelleher committed:
v0.16.0-beta4

dankelleher committed:
Gateway Plugin (#54)

* Boilerplate business-logic-free plugin for others to be based on.

* Rename Boilerplate to Gateway to create new Gateway plugin

* Some test and warning fixes

* Add gateway token validation

* Update gateway plugin key

* - Remove maxConnections
- Update IDL
- Temporary fix for decimals on voter weight

* - Remove castVote
- Remove maxVoterWeightRecord

* v0.15.2-beta10

* Gateway Plugin: Made IDL account names in the client lower-case.

* Run cargo fmt

* PR Comments and fixes

SebastianBor committed:
Publish npm 0.16.3 (#60)

* chore: publish 0.16.2 npm package

* chore: generate 0.16.3 npm package

dankelleher committed:
Merge remote-tracking branch 'upstream/master' into gateway-composable

dankelleher committed:
Civic Gateway plugin: Addressing PR comments

dankelleher committed:
Composability of gateway plugin (#62)

* Boilerplate business-logic-free plugin for others to be based on.

* Rename Boilerplate to Gateway to create new Gateway plugin

* Some test and warning fixes

* Add gateway token validation

* Update gateway plugin key

* - Remove maxConnections
- Update IDL
- Temporary fix for decimals on voter weight

* - Remove castVote
- Remove maxVoterWeightRecord

* Composability of gateway plugin

* Composability of gateway plugin - PR comments addressed - More tests TODO

* Composability of gateway plugin - Add tests for predecessor plugins, and bugfixes.

* Formatting, and removed voter weight action and target from the update voter weight instruction

* Renamed previous plugin registrar to previous plugin id

* v0.15.2-beta10

* Gateway Plugin: Made IDL account names in the client lower-case.

* Gateway Plugin: New program id supporting composability

* Gateway Plugin: New program id supporting composability

* Gateway Plugin: New program id and client lib supporting composability

* Gateway Plugin: Update Registrar instruction

* v0.16.0-beta4

* Civic Gateway plugin: Addressing PR comments

dankelleher committed:
Civic Gateway plugin: Additional PR comments

dankelleher committed:
Merge remote-tracking branch 'upstream/master' into gateway-composable
